coveo_stew/ci/runner.py

```diff
@@ -119,3 +137,141 @@ def _output_generic_report(self, environment: PythonEnvironment) -> None:
+
+            # run all other runners
+            runs.append(run := Run(self.environment, self.non_autofix_checks))
+            await run.run_and_report(parallel=self.parallel)
+
+        else:
+            runs.append(run := Run(self.environment, self.checks))
+            await run.run_and_report(parallel=self.parallel)
+
+        overall_status = get_overall_run_status(*runs)
+        echo.success(
+            f"The CI run for {self.environment.pretty_python_version} completed with status: {overall_status}"
+        )
+
+        if exceptions := [exception for run in runs for exception in run.exceptions]:
+            raise ExitWithFailure(
+                suggestions=(
+                    "If a command should be treated as a check failure, specify `check-failed-exit-codes`",
+                    "Reference: https://github.com/coveo/stew/blob/main/README.md#options)",
+                    "Try the commands in a shell to troubleshoot them faster.",
+                ),
+                failures=(
+                    f"\n------- [{runner} failed unexpectedly] -------\n\n{str(ex)}\n"
+                    for runner, ex in exceptions
+                ),
+            ) from CheckError("Unexpected errors occurred when launching external processes.")
+
+
+@dataclass
+class Run:
+    """The Run is a stateful object that runs checks and reports the results to the user."""
+
+    environment: PythonEnvironment
+    checks: Sequence[ContinuousIntegrationRunner]
+
+    @cached_property
+    def exceptions(self) -> List[Tuple[ContinuousIntegrationRunner, DetailedCalledProcessError]]:
+        """Exceptions are stored here after the run. Exceptions are cleared when `run_and_report` is called."""
+        return []
+
+    @property
+    def overall_status(self) -> RunnerStatus:
+        return get_overall_run_status(self)
+
+    async def run_and_report(
+        self, auto_fix: bool = False, feedback: bool = True, parallel: bool = True
+    ) -> None:
+        """Launch the runners and report the results to the user."""
+        self.exceptions.clear()
+
+        if auto_fix and parallel:
+            raise AssertionError(
+                "Some dev made a mistake; parallel and autofix are mutually exclusive!"
+            )
+
+        if parallel:
+            for next_result in asyncio.as_completed(
+                [runner.launch(self.environment, auto_fix=False) for runner in self.checks]
+            ):
+                self._report(await next_result, feedback=feedback)
+        else:
+            for runner in self.checks:
+                self._report(
+                    await runner.launch(self.environment, auto_fix=auto_fix), feedback=feedback
+                )
+
+    def _report(self, check: ContinuousIntegrationRunner, feedback: bool = True) -> None:
+        """Reports on a completed check."""
+        if check.status is RunnerStatus.Error:
+            self.exceptions.append((check, check.last_exception))
```

Comment:
Is there a benefit to using `@cached_property` here? I see you use it as one would any field.